opm/grid/common/ZoltanPartition.cpp

```diff
@@ -176,16 +137,239 @@ zoltanGraphPartitionGridOnRoot(const CpGrid& cpgrid,
+                                 &changes,        /* 1 if partitioning was changed, 0 otherwise */
+                                 &numGidEntries,  /* Number of integers used for a global ID */
+                                 &numLidEntries,  /* Number of integers used for a local ID */
+                                 &numImport,      /* Number of vertices to be sent to me */
+                                 &importGlobalGids,  /* Global IDs of vertices to be sent to me */
+                                 &importLocalGids,   /* Local IDs of vertices to be sent to me */
+                                 &importProcs,    /* Process rank for source of each incoming vertex */
+                                 &importToPart,   /* New partition for each incoming vertex */
+                                 &numExport,      /* Number of vertices I must send to other processes*/
+                                 &exportGlobalGids,  /* Global IDs of the vertices I must send */
+                                 &exportLocalGids,   /* Local IDs of the vertices I must send */
+                                 &exportProcs,    /* Process to which I send each of the vertices */
+                                 &exportToPart);  /* Partition to which each vertex will belong */
+
+        // This is very important: by default, zoltan sets numImport to -1,
+        // as we are only running Zoltan in serial.
+        // In order to make sense of the rest of  the code, this must be set
+        // to 0.
+        numImport = 0;
+
+
+    }
+    MPI_Barrier(cc);
+    std::vector<unsigned int> importGlobalGidsVector;
+    if (cc.rank() == root) {
+        std::vector<int> numberOfExportedVerticesPerProcess(cc.size(), 0);
+        for (int i = 0; i < numExport; ++i) {
+            numberOfExportedVerticesPerProcess[exportToPart[i]]++;
+        }
+        int dummyForRoot = 0;
+        cc.scatter<int>(numberOfExportedVerticesPerProcess.data(), &dummyForRoot, 1, root);
+
+
+        std::vector<int> offsets(cc.size() + 1, 0);
+        std::partial_sum(numberOfExportedVerticesPerProcess.begin(),
+                         numberOfExportedVerticesPerProcess.end(), offsets.begin() + 1);
+        std::vector<unsigned int> globalIndicesToSend(numExport, 0);
+        std::vector<int> currentIndex(cc.size(), 0);
+        for (int i = 0; i < numExport; ++i) {
+            if (exportToPart[i] >= currentIndex.size()) {
+                OPM_THROW(std::runtime_error, "Something wrong with Zoltan decomposition. "
+                          << "Debug information: exportToPart[i] = "
+                          << exportToPart[i] << ", " << "currentIndex.size() = " << currentIndex.size());
+            }
+            const auto index = currentIndex[exportToPart[i]]++ + offsets[exportToPart[i]];
+
+            if (index >= globalIndicesToSend.size()) {
+                OPM_THROW(std::runtime_error, "Something wrong with Zoltan decomposition. "
```

Comment:
This will create a deadlock

Reply:
True, but what is the better solution: Doing ```MPI_Abort``` here in the innermost region of the code, or having an exception catch higher up in the hiearchy that then calls ```MPI_Abort``` on an exception?

Reply:
Catching higher does not change the problem. Cleanest way is probably
```
int rc;
if(rank==0){
  rc = ...;
}
cc.scatter(rc); //pseudo code
if (rc != ZOLTAN_OK)
  thow ...
```
Would probably need more ifs.

Reply:
This would only postpone the deadlock until the next global MPI operation in the code, no?

Reply:
In my example all processes throw. But yes there need to multiple of these.

Reply:
How would all throw? Zoltan is only initialized on rank zero, all the others have not initialized Zoltan, and therefore will not get a non-zero value of  ```rc``` giving the throw.

Reply:
Nevermind, didn't see the scatter.

But why is this preferable to having

```C++
try {
   ...
} catch (...) { // or whatever exception type
   MPI_Abort(cc, -1);
}
```

this will also abort on all procs, but will have the added benifit that also non-MPI parts of the code can throw exceptions without raising a deadlock.

Reply:
You are asking the wrong person. I did that but had to change to prevent user confusion.

See  bed4093b1568068faae2ffa3151148819896ede4 for what I meant.

Reply:
Beware I did not even try to compile that.

Reply:
use 288d8a08, which compiles